external/dhcpd/dhcpd_lwip.c

```diff
@@ -103,29 +102,113 @@ typedef struct dhcp_join_data dhcp_join_data_s;
 #define DHCPD_MQ_LEN 11
 #define DHCPD_MQ_MAX_LEN 20
 
-void *_dhcpd_join_handler(void *arg)
+static int _send_mqmsg(mqd_t mq, char *msg, size_t msglen, int prio)
 {
-	dhcp_join_data_s *data = (dhcp_join_data_s *)arg;
-	int ret = ERROR;
-	int sockfd = 0;
-	struct req_lwip_data req;
+	nvdbg("dhcps send dhcps msg size %d\n", msglen);
+	while (1) {
+		int mq_ret = mq_send(mq, msg, msglen, prio);
+		if (mq_ret == -1) {
+			if (errno == EAGAIN || errno == EINTR) {
+				continue;
+			}
+			ndbg("dhcps _send_mqmsg error %d\n", errno);
+			return -1;
+		}
+		nvdbg("dhcps sent res %d sent %d\n", mq_ret, msglen);
+		break;
+	}
+	return 0;
+}
 
+static int _recv_mqmsg(mqd_t mq, char *msg, size_t msglen, int *prio)
+{
+	nvdbg("dhcps receive dhcps msg size %d\n", msglen);
+	while (1) {
+		int mq_ret = mq_receive(mq, msg, msglen, prio);
+		if (mq_ret == -1) {
+			if (errno == EAGAIN && errno == EINTR) {
+				continue;
+			}
+			ndbg("dhcps _recv_mqmsg error %d\n", errno);
+			return -1;
+		}
+		if (mq_ret == msglen) {
+			break;
+		} else {
+			ndbg("dhcps critical error %d %d\n", mq_ret, msglen);
+		}
+	}
+	return 0;
+}
+
+static void _empty_mq(mqd_t mq)
+{
+	nvdbg("dhcps empty dhcps message queue\n");
+	char msg[DHCPD_MQ_LEN];
+	struct timespec ts;
+	int status = clock_gettime(CLOCK_REALTIME, &ts);
+	if (status != 0) {
+		ndbg("dhcps ERROR clock_gettime failed\n");
+	}
+	ts.tv_nsec += 5000000; // 5 milli seconds
+	while (1) {
+		struct mq_attr mq_stat;
+		mq_getattr(mq, &mq_stat);
+		if (mq_stat.mq_curmsgs == 0) {
+			nvdbg("dhcps dhcps msg queue is empty\n");
+			return;
+		}
+		int nbytes = mq_timedreceive(mq, msg, DHCPD_MQ_LEN, 0, &ts);
+		if (nbytes == -1) {
+			if (errno == ETIMEDOUT) {
+				break;
+			} else {
+				nvdbg("dhcps empty dhcps message queue errno %d\n", errno);
+			}
+		}
+	}
+	nvdbg("dhcps empty mq\n");
+}
+
+static void _create_mq(mqd_t *mq)
+{
+	nvdbg("dhcps create queue\n");
 	struct mq_attr attr;
 	attr.mq_maxmsg = DHCPD_MQ_MAX_LEN;
 	attr.mq_msgsize = DHCPD_MQ_LEN;
 	attr.mq_flags = 0;
 	attr.mq_curmsgs = 0;
 
+	*mq = mq_open(DHCPD_MQ_NAME, O_RDWR | O_CREAT, 0666, &attr);
+	if (*mq == (mqd_t)ERROR) {
+		ndbg("dhcps Failed to open mq\n");
+	}
+	nvdbg("dhcps create mq\n");
+}
 
-	mqd_t md = mq_open(DHCPD_MQ_NAME, O_RDWR | O_CREAT, 0666, &attr);
-	if (md == (mqd_t)ERROR) {
-		ndbg("Failed to open mq\n");
-		goto go_out;
+static void _delete_mq(mqd_t mq)
+{
+	nvdbg("dhcps delete mq\n");
+	if (mq_close(mq) != 0) {
+		ndbg("dhcps fail to delete mq\n");
 	}
+}
+
+void *_dhcpd_join_handler(void *arg)
+{
+	dhcp_join_data_s *data = (dhcp_join_data_s *)arg;
+	int ret = ERROR;
+	int sockfd = 0;
+	struct req_lwip_data req;
+	mqd_t md = (mqd_t)-1;
+
+	_create_mq(&md);
+
+	_empty_mq(md);
```

Comment:
If `_create_mq()` fails, we cannot do `_empty_mq()` anymore.
So I think we'd better check the return of `_create_mq()` before calling `_empty_mq()`.

Reply:
if _create_mq() failed then system will be restarted. because without dhcp server application can't be succeeded. so I think reboot system or adding assert() would be needed.

Reply:
I agree. If the mq creation fails, intended assert is necessary.